src/R/Components/Impl/ConnectionManager/Implementation/ConnectionManager.cs

```diff
@@ -153,25 +152,71 @@ internal class ConnectionManager : IConnectionManager {
                 .ToArray();
         }
 
         private void UpdateRecentConnections() {
-            RecentConnections = new ReadOnlyCollection<IConnection>(_userConnections.Values.OrderByDescending(c => c.TimeStamp).ToList());
+            RecentConnections = new ReadOnlyCollection<IConnection>(_userConnections.Values.OrderByDescending(c => c.LastUsed).ToList());
             SaveConnectionsToSettings();
             RecentConnectionsChanged?.Invoke(this, new EventArgs());
         }
 
+        private Dictionary<Uri, IConnection> CreateConnectionList() {
+            var connections = GetConnectionsFromSettings();
+            var localEngines = new RInstallation().GetCompatibleEngines();
+            if (connections.Count == 0) {
+                if (!localEngines.Any()) {
+                    var message = string.Format(CultureInfo.InvariantCulture, Resources.NoLocalR, Environment.NewLine + Environment.NewLine, Environment.NewLine);
+                    if (_shell.ShowMessage(message, MessageButtons.YesNo) == MessageButtons.Yes) {
+                        var installer = _shell.ExportProvider.GetExportedValue<IMicrosoftRClientInstaller>();
+                        installer.LaunchRClientSetup(_shell);
+                        return connections;
+                    }
+                }
+                // No connections, may be first use or connections were somehow removed.
+                // Add local connections so there is at least something available.
+                foreach (var e in localEngines) {
+                    var c = CreateConnection(e.Name, e.InstallPath, string.Empty, isUserCreated: false);
+                    connections[new Uri(e.InstallPath, UriKind.Absolute)] = c;
+                }
+            } else {
+                // Remove missing engines and add engines missing from saved connections
+                // Set 'is used created' to false if path points to locally found interpreter
+                foreach (var kvp in connections.ToList()) {
+                    if (!kvp.Value.IsRemote) {
+                        bool valid = false;
+                        try {
+                            var info = new RInterpreterInfo(kvp.Value.Name, kvp.Value.Path);
+                            valid = info.VerifyInstallation();
+                        } catch (Exception ex) when (!ex.IsCriticalException()) {
+                            GeneralLog.Write(ex);
+                        }
+                        if (!valid) {
+                            connections.Remove(kvp.Key);
+                        }
+                    }
+                }
+
+                // Add newly installed engines
+                foreach (var e in localEngines) {
+                    if (!connections.Values.Any(x => x.Path.TrimTrailingSlash().EqualsIgnoreCase(e.InstallPath.TrimTrailingSlash()))) {
```

Comment:
We can add a comparison method to the `IOExtensions` and then change dictionary comparer:
```{csharp}
public static CompareDirectoryPaths(this string path, string otherPath) {
    var pathLength = GetTrailingSlashTrimmedPathLength(path);
    var otherPathLength = GetTrailingSlashTrimmedPathLength(otherPath);
      
    return pathLength == otherPathLength && (pathLength == 0 || string.Compare (path, 0, otherPath, 0, otherPathLength, true, CultureInfo.InvariantCulture) == 0);
}

private static int GetTrailingSlashTrimmedPathLength(string path) {
    if (string.IsNullOrEmpty(path)) {
        return 0;
    } 
    if (path.EndsWithOrdinal("\\") || path.EndsWithOrdinal("/"))) {
        return path.Length - 1;
    }
    return path.Length;
}
```

---
In reply to: [77673244](https://github.com/Microsoft/RTVS/pull/2246#discussion_r77673244) [](ancestors = 77673244)